### ISSUE TITLE
Explicit cast should warn if nested nullability does not match

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#29641]
The second conversion, from `List<string>` to `List<string?>` should result in a warning.
```c#
using System.Collections.Generic;
using System.Runtime.CompilerServices;
[NonNullTypes]
class C
{
    static void F(List<string> x)
    {
        var y = (List<object>)x;  // error CS0030: Cannot convert 'List<string>' to 'List<object>'
        var z = (List<string?>)x; // no warning
    }
}
```
